underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -0,0 +1,406 @@
+    // This point is only reached if the operation failed more
+    // than the allowed retry count
+    throw new IOException("Unable to init multipart upload to " + mKey, lastException);
+  }
+
+  /**
+   * Creates a new temp file to write to.
+   */
+  private void initNewFile() throws IOException {
+    mFile = new File(PathUtils.concatPath(CommonUtils.getTmpDir(), UUID.randomUUID()));
+    if (mHash != null) {
+      mLocalOutputStream =
+          new BufferedOutputStream(new DigestOutputStream(new FileOutputStream(mFile), mHash));
+    } else {
+      mLocalOutputStream = new BufferedOutputStream(new FileOutputStream(mFile));
+    }
+    mOffset = 0;
+    LOG.debug("Init new temp file @ {}", mFile.getPath());
+  }
+
+  /**
+   * Uploads part async.
+   */
+  private void uploadPart() throws IOException {
+    if (mFile == null) {
+      return;
+    }
+    mLocalOutputStream.close();
+    int partNumber = mPartNumber.getAndIncrement();
+    File newFileToUpload = new File(mFile.getPath());
+    mFile = null;
+    mLocalOutputStream = null;
+    UploadPartRequest uploadRequest = new UploadPartRequest()
+        .withBucketName(mBucketName)
+        .withKey(mKey)
+        .withUploadId(mUploadId)
+        .withPartNumber(partNumber)
+        .withFile(newFileToUpload)
+        .withPartSize(newFileToUpload.length());
+    execUpload(uploadRequest, newFileToUpload, partNumber);
+  }
+
+  /**
+   * Executes the upload part request.
+   */
+  private void execUpload(UploadPartRequest request, File file, int partNumber) {
+    ListenableFuture<PartETag> futureTag =
+        mExecutor.submit((Callable) () -> {
+          PartETag partETag;
+          AmazonClientException lastException;
+          try {
+            do {
+              try {
+                partETag = mClient.uploadPart(request).getPartETag();
+                return partETag;
+              } catch (AmazonClientException e) {
+                lastException = e;
+              }
+            } while (mRetryPolicy.attempt());
+          } finally {
+            // Delete the uploaded or failed to upload file
+            if (!file.delete()) {
+              LOG.error("Failed to delete temporary file @ {}", file.getPath());
+            }
+          }
+          throw new IOException("Fail to upload part " + partNumber + " to " + mKey,
+              lastException);
+        });
+    mFutureTags.add(futureTag);
+    LOG.debug("Submit upload part request. key={}, partNum={}, file={}, fileSize={}.",
+        mKey, partNumber, file.getPath(), file.length());
+  }
+
+  /**
+   * Waits for the submitted upload tasks to finish.
+   */
+  private void waitForAllPartsUpload() throws IOException {
+    try {
+      mTags.addAll(Futures.allAsList(mFutureTags).get());
+    } catch (ExecutionException e) {
+      // No recover ways so that we need to cancel all the upload tasks
+      // and abort the multipart upload
+      for (ListenableFuture<PartETag> future : mFutureTags) {
```

Comment:
cancel the future returned by `addAll` will cancel all futures in mFutureTags.

Reply:
Thanks! Great and convenient methods.

Changed to
```
Futures.allAsList(mFutureTags).cancel(true);
```